service/internal/checksum/middleware_compute_input_checksum_test.go

```diff
@@ -968,6 +968,246 @@ func TestComputeInputPayloadChecksum(t *testing.T) {
 	}
 }
 
+func TestComputeInputPayloadChecksumRetry(t *testing.T) {
+	cases := map[string]struct {
+		initContext       func(context.Context) context.Context
+		buildInput        middleware.BuildInput
+		updatedStreamBody string
+
+		expectHeader           http.Header
+		expectPayload          []byte
+		expectDeferToFinalize  bool
+		expectChecksumMetadata map[string]string
+		expectErr              string
+	}{
+		"http request with stream unchanged": {
+			initContext: func(ctx context.Context) context.Context {
+				return internalcontext.SetChecksumInputAlgorithm(ctx, string(AlgorithmCRC32))
+			},
+			buildInput: middleware.BuildInput{
+				Request: func() *smithyhttp.Request {
+					r := smithyhttp.NewStackRequest().(*smithyhttp.Request)
+					r.URL, _ = url.Parse("http://example.aws")
+					r.ContentLength = 11
+					r = requestMust(r.SetStream(bytes.NewReader([]byte("hello world"))))
+					return r
+				}(),
+			},
+			expectHeader: http.Header{
+				"X-Amz-Checksum-Crc32": []string{"DUoRhQ=="},
+			},
+			expectPayload: []byte("hello world"),
+			expectChecksumMetadata: map[string]string{
+				"CRC32": "DUoRhQ==",
+			},
+			expectErr: "mock error in first attempt",
+		},
+		"http request with stream updated": {
+			initContext: func(ctx context.Context) context.Context {
+				return internalcontext.SetChecksumInputAlgorithm(ctx, string(AlgorithmCRC32))
+			},
+			buildInput: middleware.BuildInput{
+				Request: func() *smithyhttp.Request {
+					r := smithyhttp.NewStackRequest().(*smithyhttp.Request)
+					r.URL, _ = url.Parse("http://example.aws")
+					r.ContentLength = 11
+					r = requestMust(r.SetStream(bytes.NewReader([]byte("hello world"))))
+					return r
+				}(),
+			},
+			updatedStreamBody: "world hello",
+			expectHeader: http.Header{
+				"X-Amz-Checksum-Crc32": []string{"DUoRhQ=="},
```

Comment:
> guarantees payload consistency check in case request body is updated between reattempts.

is that right? it seems like we're keeping the same checksum even with updated payload

Reply:
The logic here is "if there is a checksum value available from previous failed request, just use it", so the old checksum should still be set as header